lib/active_storage/blob_io.rb

```diff
@@ -0,0 +1,50 @@
+# Acts as a proxy to turn ActiveStorage file into IO object
+
+module FormatParser
+  module ActiveStorage
+    class BlobIO
+      # @param blob[ActiveStorage::Blob] the file with linked service
+      # @return [BlobIO]
+      def initialize(blob)
+        @blob = blob
+        @service = blob.service
+        @pos = 0
+      end
+
+      # Emulates IO#read, but requires the number of bytes to read.
+      # Rely on `ActiveStorage::Service.download_chunk` of each hosting type (local, S3, Azure, etc)
+      #
+      # @param n_bytes[Integer] how many bytes to read
+      # @return [String] the read bytes
+      def read(n_bytes)
+        http_range = (@pos..(@pos + n_bytes))
```

Comment:
The reason is that HTTP ranges are fundamentally exclusive. When you want to request 1 byte at offset 0, you do it like so:

`Range: bytes=0-0`

The `read()` method receives 1 as argument, so if you do not subtract 1 you will do this:

`http_range = @pos..(@pos + 1) = 0..1

and then perform a request with

`Range: bytes=0-1`

because the range `begin` and `end` values get templated into the header. But requesting `0-1` will not give you 1 byte but will give you two bytes instead.

Reply:
It might be that `download_chunk` in this case does it for you (!) that's why it worked out of the box here. Can you please double-check?

Reply:
Thanks @julik 
- disk service is pretty basic https://github.com/rails/rails/blob/master/activestorage/lib/active_storage/service/disk_service.rb#L43-L44
- s3 service and azure do something different. `object_for(key).get(range: "bytes=#{range.begin}-#{range.exclude_end? ? range.end - 1 : range.end}")`
- GCS service use a basic `download` from the `google-cloud-storage` gem. From [the gem doc](https://googleapis.dev/ruby/google-cloud-storage/latest/Google/Cloud/Storage/File.html#download-instance_method): 
    >  range (Range, String) (defaults to: nil) — Optional. The byte range of the file's contents to download or a string header value. Provide this to perform a partial download. When a range is provided, no verification is performed regardless of the verify parameter's value.
While digging deeper for GCS I do not see something specific. https://github.com/googleapis/google-cloud-ruby/blob/e4bcb6ec8b8b525363b4add6f45516256aa33489/google-cloud-storage/lib/google/cloud/storage/service.rb#L568

I tried with and without `-1` and didn't see any difference. Is there a specific scenario where it can be an issue?